File: apps/api/src/modules/auth/auth.controller.ts
import { Body, Controller, Get, Patch, Post, Query, UseGuards } from '@nestjs/common';
import { AuthService } from './auth.service';
import { FirebaseAuthGuard } from '../../common/guards/firebase-auth.guard';
import { CurrentUser, CurrentUserPayload } from '../../common/decorators/current-user.decorator';
import { RegisterDto } from './dto/register.dto';
import { RegisterFcmDto } from './dto/register-fcm.dto';
import { Public } from '../../common/decorators/public.decorator';

@Controller('auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  /** Get current user profile */
  @Get('me')
  @UseGuards(FirebaseAuthGuard)
  async getMe(@CurrentUser() user: CurrentUserPayload) {
    return this.authService.getMe(user.id);
  }

  /** Update auth profile (email) from profile settings */
  @Patch('me')
  @UseGuards(FirebaseAuthGuard)
  async updateMe(
    @CurrentUser() user: CurrentUserPayload,
    @Body() body: { email?: string },
  ) {
    return this.authService.updateProfile(user.id, { email: body.email });
  }

  /** Complete profile after OTP login (register page) */
  @Post('register')
  @UseGuards(FirebaseAuthGuard)
  async register(
    @CurrentUser() user: CurrentUserPayload,
    @Body() body: { name?: string; email?: string; password?: string; role?: string },
  ) {
    return this.authService.updateProfile(user.id, body);
  }

  @Post('register-fcm')
  @UseGuards(FirebaseAuthGuard)
  async registerFcmToken(
    @CurrentUser() user: CurrentUserPayload,
    @Body() dto: RegisterFcmDto,
  ) {
    return this.authService.registerFcmToken(user.id, dto);
  }

  /** Verify Firebase ID token — used for web app session init */
  @Post('verify-token')
  @Public()
  async verifyToken(@Body('idToken') idToken: string) {
    return this.authService.verifyAndGetOrCreateUser(idToken);
  }

  /** Check if phone is a test account (bypass Firebase OTP) */
  @Get('is-test-phone')
  @Public()
  async isTestPhone(@Query('phone') phone: string) {
    return this.authService.isTestPhone(phone);
  }

  /** Step 1: Send OTP to phone */
  @Post('otp/send')
  @Public()
  async sendOtp(@Body('phone') phone: string) {
    return this.authService.sendOtp(phone);
  }

  /** Step 2: Verify OTP → get session token */
  @Post('otp/verify')
  @Public()
  async verifyOtp(
    @Body('phone') phone: string,
    @Body('otp') otp: string,
  ) {
    return this.authService.verifyOtp(phone, otp);
  }

  /** Email + password login */
  @Post('login')
  @Public()
  async loginEmail(
    @Body('email') email: string,
    @Body('password') password: string,
  ) {
    return this.authService.loginEmail(email, password);
  }

  /** Facebook social login */
  @Post('social/facebook')
  @Public()
  async socialFacebook(@Body('idToken') idToken: string) {
    return this.authService.socialFacebook(idToken);
  }

  /** Link phone to Facebook account (after phone OTP verification) */
  @Post('social/link-phone')
  @UseGuards(FirebaseAuthGuard)
  async linkPhone(
    @CurrentUser() user: CurrentUserPayload,
    @Body('phoneIdToken') phoneIdToken: string,
  ) {
    await this.authService.linkPhone(user.id, phoneIdToken);
    return { success: true };
  }

  /** Logout — revoke Firebase tokens */
  @Post('logout')
  @UseGuards(FirebaseAuthGuard)
  async logout(@CurrentUser() user: CurrentUserPayload) {
    await this.authService.logout(user.firebaseUid);
    return { success: true };
  }
}

File: apps/api-kotlin/src/main/kotlin/vn/gada/api/auth/AuthController.kt
package vn.gada.api.auth

import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.*
import vn.gada.api.common.exception.UnauthorizedException
import vn.gada.api.common.security.AuthUser

@RestController
@RequestMapping("/auth")
class AuthController(private val authService: AuthService) {

    /** GET /auth/me — Get current user profile */
    @GetMapping("/me")
    fun getMe(@AuthenticationPrincipal user: AuthUser?): ResponseEntity<Map<String, Any?>> {
        if (user == null) throw UnauthorizedException("Unauthorized")
        val profile = authService.getMe(user.id)
        return ok(profile)
    }

    /** POST /auth/register — Complete profile after OTP login */
    @PostMapping("/register")
    fun register(
        @AuthenticationPrincipal user: AuthUser?,
        @RequestBody body: Map<String, Any?>
    ): ResponseEntity<Map<String, Any?>> {
        if (user == null) throw UnauthorizedException("Unauthorized")
        val name = body["name"] as? String
        val email = body["email"] as? String
        val result = authService.updateProfile(user.id, name, email)
        return ok(result)
    }

    /** POST /auth/register-fcm — Register FCM token */
    @PostMapping("/register-fcm")
    fun registerFcm(
        @AuthenticationPrincipal user: AuthUser?,
        @RequestBody body: Map<String, Any?>
    ): ResponseEntity<Map<String, Any?>> {
        if (user == null) throw UnauthorizedException("Unauthorized")
        val token = body["token"] as? String ?: throw vn.gada.api.common.exception.BadRequestException("token is required")
        val platform = body["platform"] as? String ?: "android"
        val result = authService.registerFcmToken(user.id, token, platform)
        return ok(result)
    }

    /** POST /auth/verify-token — Verify Firebase ID token (web session init) */
    @PostMapping("/verify-token")
    fun verifyToken(@RequestBody body: Map<String, Any?>): ResponseEntity<Map<String, Any?>> {
        val idToken = body["idToken"] as? String ?: throw vn.gada.api.common.exception.BadRequestException("idToken is required")
        val result = authService.verifyAndGetOrCreateUser(idToken)
        return ok(result)
    }

    /** POST /auth/otp/send — Send OTP to phone */
    @PostMapping("/otp/send")
    fun sendOtp(@RequestBody body: Map<String, Any?>): ResponseEntity<Map<String, Any?>> {
        val phone = body["phone"] as? String ?: throw vn.gada.api.common.exception.BadRequestException("phone is required")
        val result = authService.sendOtp(phone)
        return ok(result)
    }

    /** POST /auth/otp/verify — Verify OTP and get session token */
    @PostMapping("/otp/verify")
    fun verifyOtp(@RequestBody body: Map<String, Any?>): ResponseEntity<Map<String, Any?>> {
        val phone = body["phone"] as? String ?: throw vn.gada.api.common.exception.BadRequestException("phone is required")
        val otp = body["otp"] as? String ?: throw vn.gada.api.common.exception.BadRequestException("otp is required")
        val result = authService.verifyOtp(phone, otp)
        return ok(result)
    }

    /** POST /auth/login — Email + password login */
    @PostMapping("/login")
    fun loginEmail(@RequestBody body: Map<String, Any?>): ResponseEntity<Map<String, Any?>> {
        val email = body["email"] as? String ?: throw vn.gada.api.common.exception.BadRequestException("email is required")
        val password = body["password"] as? String ?: throw vn.gada.api.common.exception.BadRequestException("password is required")
        val result = authService.loginEmail(email, password)
        return ok(result)
    }

    /** POST /auth/social/facebook — Facebook social login */
    @PostMapping("/social/facebook")
    fun socialFacebook(@RequestBody body: Map<String, Any?>): ResponseEntity<Map<String, Any?>> {
        val idToken = body["idToken"] as? String ?: throw vn.gada.api.common.exception.BadRequestException("idToken is required")
        val result = authService.socialFacebook(idToken)
        return ok(result)
    }

    /** POST /auth/social/google — Google social login (same flow as /verify-token) */
    @PostMapping("/social/google")
    fun socialGoogle(@RequestBody body: Map<String, Any?>): ResponseEntity<Map<String, Any?>> {
        val idToken = body["idToken"] as? String ?: throw vn.gada.api.common.exception.BadRequestException("idToken is required")
        val result = authService.verifyAndGetOrCreateUser(idToken)
        return ok(result)
    }

    /** POST /auth/logout — Revoke Firebase tokens */
    @PostMapping("/logout")
    fun logout(@AuthenticationPrincipal user: AuthUser?): ResponseEntity<Map<String, Any?>> {
        if (user == null) throw UnauthorizedException("Unauthorized")
        authService.logout(user.firebaseUid)
        return ok(mapOf("success" to true))
    }

    private fun ok(data: Any?): ResponseEntity<Map<String, Any?>> =
        ResponseEntity.ok(mapOf("statusCode" to 200, "data" to data))
}
